File: viewer/src/main/java/com/tianxing/hotflyer/viewer/network/provider/DownloadLinkProvider.java
package com.tianxing.hotflyer.viewer.network.provider;

import java.util.List;

import com.tianxing.hotflyer.viewer.adapter.item.DownloadLink;
import com.tianxing.hotflyer.viewer.adapter.item.MagnetLink;

import okhttp3.ResponseBody;
import retrofit2.Call;

/**
 * Project: JAViewer
 */
public abstract class DownloadLinkProvider {
    public abstract Call<ResponseBody> search(String keyword, int page);

    public abstract List<DownloadLink> parseDownloadLinks(String htmlContent);

    public abstract Call<ResponseBody> get(String url);

    public abstract MagnetLink parseMagnetLink(String htmlContent);

    public static DownloadLinkProvider getProvider(String name) {
        switch (name.toLowerCase().trim()) {
            case "btso":
                return new BTSOLinkProvider();
            case "torrentkitty":
                return new TorrentKittyLinkProvider();
            default:
                return null;
        }
    }
}

File: viewer/src/main/java/com/tianxing/hotflyer/viewer/network/provider/DownloadLinkProvider.kt
package com.tianxing.hotflyer.viewer.network.provider

import com.tianxing.hotflyer.viewer.adapter.item.DownloadLink
import com.tianxing.hotflyer.viewer.adapter.item.MagnetLink

import okhttp3.ResponseBody
import retrofit2.Call

/**
 * Project: JAViewer
 */
abstract class DownloadLinkProvider {
    abstract fun search(keyword: String, page: Int): Call<ResponseBody>?

    abstract fun parseDownloadLinks(htmlContent: String): List<DownloadLink>

    abstract operator fun get(url: String): Call<ResponseBody>?

    abstract fun parseMagnetLink(htmlContent: String): MagnetLink?

    companion object {

        fun getProvider(name: String): DownloadLinkProvider? {
            when (name.toLowerCase().trim { it <= ' ' }) {
                "btso" -> return BTSOLinkProvider()
                "torrentkitty" -> return TorrentKittyLinkProvider()
                else -> return null
            }
        }
    }
}
